Update `LinkControl` component to utilitse dynamic settings for additional settings "drawer"

getdave opened the pull request:
Closes https://github.com/WordPress/gutenberg/issues/18206.

The new experimental Link UI provided by `LinkControl` allows for a currently selected link to have settings such as "Open in new tab"...etc

As things stand the only setting that is possible is "new tab" as this has been hard coded.

This PR updates so that multiple settings can be provided by passing the `currentSettings` prop to `LinkControl`:

```jsx
<LinkControl
	// ...other props here
	currentSettings={ [
		{
			id: 'newTab',
			title: 'Open in New Tab',
			checked: false,
		},
		{ // this is custom!
			id: 'noFollow',
			title: 'No follow',
			checked: true,
		},
	] }
/>

```

## How has this been tested?
Automated tests cover this change.

## Types of changes
Breaking change (fix or feature that would cause existing functionality to not work as expected).

As this updates `LinkControl` we can expect [this PR](https://github.com/WordPress/gutenberg/pull/18062) to break. It will need updating so that:

* `new-tab` becomes `newTab` in any settings provided.


## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

getdave committed:
Adds basic coverage test for new tab setting

getdave committed:
Refactor ‘new-tab` to camelCase for consistency

getdave committed:
Fix bug with null render if settings has a length and convert to array

Make settings an array so that it can be ordered. Fix incorrect conditional testing for .length to ensure it passes if settings _does_ have a length.

getdave committed:
Updates to loop over supplied settings

Previously only a “new tab” setting was hardcoded. Update so retain “new tab” as the default, but also allow for custom settings if/when provided.

getdave commented:
@phpbits This is relevant to https://github.com/WordPress/gutenberg/issues/17557#issuecomment-540050846

getdave committed:
Updates docs

getdave commented:
@retrofox I'm interested in how you see my change working with https://github.com/WordPress/gutenberg/pull/18062/files#diff-17e2993cb0b8cf64434015038b235162R214

retrofox reviewed (approved):
### Testing

I've tested locally in my dev env. It looks pretty good!

![link-control-dynamic](https://user-images.githubusercontent.com/77539/68240292-647fb800-ffeb-11e9-9efe-cca6214c1ed7.gif)

It's pretty easy to use and narrows very good with the block attributes:

```es6
<LinkControl
	currentSettings={ [
		{
			id: 'opensInNewTab',
			title: __( 'Open in New Tab' ),
			checked: opensInNewTab,
		},
		{
			id: 'noFollow',
			title: 'No follow',
			checked: noFollow,
		},
	] }
	onSettingsChange={ ( setting, value ) => setAttributes( { [ setting ]: value } ) }
/>
```

retrofox commented:
btw feel free to go-ahead not waiting for https://github.com/WordPress/gutenberg/pull/18062 👍

getdave commented:
I'll be rebasing this branch shortly and tweaking the visual styles. Then we're 👍 to merge.

getdave committed:
Updates Nav Item Block to utilise new setting format

getdave committed:
Updates spacing between individual setting controls

getdave commented:
@retrofox Having to pass the entire config object even to use the default setting feels fairly laboirious

```
<LinkControl
	currentSettings={ [
		{
			id: 'opensInNewTab', // why can't I omit this?
			title: __( 'Open in New Tab' ), // why can't I omit this?
			checked: opensInNewTab, 
		},
	] }
	onSettingsChange={ ( setting, value ) => setAttributes( { [ setting ]: value } ) }
```

Is there a way to make this API nicer?

getdave committed:
Update docs for settings

retrofox commented:
While I was testing I immediately thought that the `id` and `checked` properties could be combined in just one. At least it could be optional:
If the `checked` is not defined we can try to apply the value through of the `id`

```es6
<LinkControl
	currentSettings={ [
		{
			id: 'opensInNewTab',
			title: __( 'Open in New Tab' ), // why can't I omit this?
		},
	] }
```

About the title, maybe? I guess that maybe it could be handled by the parent component? Not sure, though. We can just remove the title if it isn't defined.

retrofox commented:
Dave, I think we should update how the [Menu Item updates the link settings](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/navigation-menu-item/edit.js#L48-L50): something like that?

```es6
const updateLinkSetting = ( setter ) => ( setting, value ) => {
	setter( { [setting]: value } );
};
```

retrofox commented:
Also, we will need to update the [setting ID here](https://github.com/WordPress/gutenberg/pull/18285/files#diff-17e2993cb0b8cf64434015038b235162R222) to `opensInNewTab`

getdave committed:
Updates setting key/id to match attribute to simply and improve consistency

getdave commented:
Ok @retrofox this is updated.

retrofox commented on packages/block-library/src/navigation-menu-item/edit.js:
```diff
@@ -45,9 +45,7 @@ import { Fragment, useState, useEffect } from '@wordpress/element';
-	if ( setting === 'newTab' ) {
-		setter( { opensInNewTab: value } );
-	}
+	setter( { opensInNewTab: value } );
```
why not pick up the setting ID to update from the callback function instead of setting it explicitly? 

```es6
setter( { [ setting ]: value } );
```

So if we add another setting in the future such as `nofollow` for instance, we won't need to concern about propagating these values

getdave replied:
That's a copy/paste mistake by me! Sorry

getdave replied:
Fixed

getdave committed:
Fix Nav Menu Block dynamic setting of attribute

Addresses https://github.com/WordPress/gutenberg/pull/18285#discussion_r343152809
